tests/README.md

```diff
@@ -0,0 +1,1510 @@
+
+<a name="execute-ilias-installation-bound-tests"></a>
+#### Execute ILIAS installation bound tests
+```
+./libs/composer/vendor/bin/phpunit -c ./Services/PHPUnit/config/PhpUnitConfig.xml \
+	--colors=always \
+	--no-globals-backup \
+	--report-useless-tests \
+	--disallow-todo-tests \
+	--group needsInstalledILIAS
+```
+
+<a name="execute-all-tests"></a>
+#### Execute all tests
+```
+./libs/composer/vendor/bin/phpunit -c ./Services/PHPUnit/config/PhpUnitConfig.xml \
+	--colors=always \
+	--report-useless-tests \
+	--disallow-todo-tests \
+	--no-globals-backup
+```
+
+<a name="execute-only-installation-unbound-tests"></a>
+#### Execute only installation unbound tests
+```
+./libs/composer/vendor/bin/phpunit -c ./Services/PHPUnit/config/PhpUnitConfig.xml \
+	--colors=always \
+	--no-globals-backup \
+	--report-useless-tests \
+	--disallow-todo-tests \
+	--exclude-group needsInstalledILIAS
+```
+
+<a name="guidelines"></a>
+## Guidelines
+
+<a name="foreword"></a>
+### Foreword
+May the reader is asking him self why this guidelines refer to a book
+which writes about JUnit testing in Java 8.
+
+The reason for this is that unit testing in his very nature is the same in
+every language. The second reason is that no recently written books are
+on the market which describe modern unit testing with PHPUnit.
+
+<a name="naming"></a>
+### Naming
+"Rework test names and code to tell stories." (Langr 2015)
+This means to treat the tests as a specification which tells everything
+about the behaviours of a class.
+
+<a name="class"></a>
+#### Class
+The filename of the test class should always be named like *\<class name of the implementation\>Test.php*.
+Furthermore, the test class should always be named as the class which is tested by the unit test class.
+For example the real class is called *Car* the corresponding test class would be named *CarTest* and the filename
+ *CarTest.php*.
+
+<a name="unit-test-method"></a>
+#### Unit-Test (Method)
+The method name must describe what your test is doing. For example the name "testSomeBasics" is not really saying
+much about the test. It is also possible that the test is actually testing multiple behaviours because of the
+generic name.
+
+Some good, more descriptive names have the following forms:
+- doingSomeOperationGeneratesSomeResult
+- someResultOccursUnderSomeCondition
+- whenDoingSomeBehaviourThenSomeResultOccurs
+
+<a name="current-state"></a>
+##### Current state
+The style how methods are named in the tests are different the most common two
+are lower camel case and snake case.
+
+<a name="proposal"></a>
+###### Proposal
+A more or less recent study has shown that *snake_case* and *CamelCase* provide the same readability.
+However the new class names in ILIAS are *CamelCase* and the old ones *lowerCamelCase*.
+Therefore, the new test methods should be written in lower camel case to match the
+camel case class names in a more consistent way.
```

Comment:
@klees I don't like that a lot. However, this seems like a matter of taste. I can live with whatever is applied consistently. What do the others prefer?

Reply:
I guess it really depends which style a developer is used to.
Is there already a code style guide which describes all the conventions ?

Reply:
If this was just a matter of taste or style, we could as well find a convention for that. My argument is exactly opposite: I see value in underscore_case for test methods if the main codebase uses DromedarCase because it visually allows to identify 
`NamesFromTheCodebase` in `names_of_testcases_over_NamesFromTheCodebase` and thus disambiguates descriptive testcase names. I mean, compare `doingSomeOperationGeneratesSomeResult`, `doing_SomeOperation_generates_SomeResult` and `doing_SomeOperation_generates_some_Result`, ...

I could also live with `doingSomeOperationGeneratesSomeResult`, though.

Reply:
@klees If I get you right, you would like to have a consistent naming of the implementation and the test methods ?
For example:
`Car::startEngine()`and `CarTest::testStartEngineWhichShouldSucceed()`
or with snake case
`Car::start_engine()`and `CarTest::test_start_engine_which_should_succeed()`
but not
`Car::start_engine()`and `CarTest::testStartEngineWhichShouldSucceed()`

Reply:
Nope, that's @Amstutz, I guess =)

I would propose: `Car::startEngine()` and `CarTest::test_startEngine_which_should_succeed()` because then I can clearly see the name of the method `startEngine` in the name of my test method. So my thesis is, that it is actually good to have different naming schemes in the code base and the tests thereof.

Reply:
@klees 
> Nope, that's @Amstutz, I guess =)

Where did you get that from? All I am saying is, that whatever we chose (I still have a slight preference for the lowerCamelCase, we should apply consistently in all our test cases across ilias.

Reply:
@Amstutz Just guessing. So in general you are ok with different naming conventions for tests and the rest of the codebase?

Reply:
Not a fan of it, but well, I can live with it. So to the point: yes, as long as we apply it consistently through our test functions.

Reply:
Ok. So next question could be: Does my argument about identification and disambiguation of names make any sense to you?

Reply:
I do not really see the need of this hard distinction, since those names are mostly only used in the methods declarations and used by phpunit. Further there is already the test prefix. On the other hand I do see the potential of lots of devs. that will break with the convention, since they are used to the camel case version in the rest of the code. So there will be some house keeping to do, which I am unsure if it is worth the effort. But as said, I can live with it, and I am also willing to help with the house keeping chore if we would decide for that.

Reply:
In my opinion we should stick to one coding style because test are an important part of the codebase and should not really differ from the style of the implementation. Furthermore it could complicate the usage of static analysis tools which report for example code style violations.

Reply:
@d3r1w +1 for your point concerning the more complicate use of tools that perform code analysis.

Reply:
Ok, I would prefer the (imo) better readability, but I'm fine with that too.